ethcore/vm/src/error.rs

```diff
@@ -46,7 +46,7 @@ pub enum Error {
 	/// to position that wasn't marked with JUMPDEST instruction
 	BadJumpDestination {
 		/// Position the code tried to jump to.
-		destination: usize
+		destination: U256
```

Comment:
I understand that `usize` is a lossy type here, but it could have been done for performance reasons, please checkout benches in `evm` to make sure it doesn't introduce a regression:
```
# on this branch
cd ethcore/evm
cargo bench
# and then on master
git checkout master
cargo bench
```

Reply:
I changed this cause the `Failure 1: invalid jump` test in https://eips.ethereum.org/EIPS/eip-2315, the error should be `BadJumpDestination(0x01000000000000000c)` instead `BadJumpDestination(0xc)`

Reply:
Maybe this is related in some way to EIP-1985: Sane limits for certain EVM parameters. @wesrer, any thoughts on this?

Reply:
After running the benches locally, I see a couple of regressions locally
```
simple_loop_log0_usize  time:   [1.0207 us 1.0276 us 1.0358 us]                                    
                        change: [+12.938% +13.869% +14.978%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 18 outliers among 100 measurements (18.00%)
  7 (7.00%) low mild
  4 (4.00%) high mild
  7 (7.00%) high severe

simple_loop_log0_u256   time:   [1.2721 us 1.2801 us 1.2901 us]                                   
                        change: [+8.4202% +9.5204% +10.547%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 14 outliers among 100 measurements (14.00%)
  2 (2.00%) low mild
  5 (5.00%) high mild
  7 (7.00%) high severe
```
although the measurements here are noisy as you can see from the number of outliers, could you confirm this?

Reply:
I got these ±changes (outliers):

| test                                               | 1st run | 2nd run | 3rd run | 4rd run
|---------------------------------------------|----------|-------------|------------|----------|
| simple_loop_log0_usize  master  | -7.5 (2)   |   +1.0 (12)      |   +0.3 (1)   |  +5.1 (11)  |
| simple_loop_log0_u256  master  |    -10.0 (3)   |   +3.1 (1)       |  -3.8 (3)     | +10.2 (6)     |
| simple_loop_log0_usize  branch  |  +11.1 (10)   |  -1.4 (2)      |  -0.8 (11)      | +0.5 (4)    | 
| simple_loop_log0_u256  branch  |   -4.7 (6) | +4.1 (13)       | -2.2 (8)        | 1.3 (26)     |

I do not see any significative change there, but I do not understand this variability, 
I got with `openssl speed aes` on 16 size blocks 18602897, 19709049, 19716673 & 19714374 benches that are a lot more stable.

I suppose, that I need to learn more about these bench tests.

Reply:
the benches use https://github.com/bheisler/criterion.rs, it saves the results of a run and when you run it again, it displays changes since the previous run, the results in my comment are from running on master first and then on your branch, I don't know if you've done the same

Reply:
ok, I tried to get more reliable results but it was not the proper way, lol.

first running on eip branch and next to master the result is:

```
simple_loop_log0_usize  time:   [1.3510 us 1.3517 us 1.3524 us]
                        change: [-11.111% -10.978% -10.841%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) low severe
  5 (5.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe

Benchmarking simple_loop_log0_u256: Collecting 100 samples in estimated 5.0054 s (3.0M iterati                                                                                              simple_loop_log0_u256   time:   [1.7443 us 1.7455 us 1.7467 us]
                        change: [-13.154% -13.022% -12.887%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) low mild
  4 (4.00%) high mild
  2 (2.00%) high severe
```

backward is

```
Benchmarking simple_loop_log0_usize: Collecting 100 samples in estimated 5.0079 s (3.2M iterat                                                                                              simple_loop_log0_usize  time:   [1.5556 us 1.5569 us 1.5580 us]
                        change: [+11.043% +11.493% +11.942%] (p = 0.00 < 0.05)
                        Performance has regressed.

Benchmarking simple_loop_log0_u256: Collecting 100 samples in estimated 5.0023 s (2.5M iterati                                                                                              simple_loop_log0_u256   time:   [2.0294 us 2.0306 us 2.0319 us]
                        change: [+21.838% +22.017% +22.206%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe
```

yes, there is a performance regression there. I was not aware about this benches, it seems that this must be in our CI pipeline (cc/ @vorot93)

maybe boxing the U256 could help?

Reply:
I can confirm the regression (and the noisiness):

```
$ cargo bench --bench basic simple_loop_log0_ -- --baseline eip-2315
    Finished bench [optimized] target(s) in 0.09s
     Running /home/dvdplm/work/openethereum/target/release/deps/basic-1f7466fb8393ff63
Gnuplot not found, using plotters backend
simple_loop_log0_usize  time:   [945.80 ns 947.68 ns 949.57 ns]                                    
                        change: [-10.173% -9.9695% -9.7719%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

simple_loop_log0_u256   time:   [1.1972 us 1.1979 us 1.1986 us]                                   
                        change: [-13.667% -13.493% -13.326%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild
  1 (1.00%) high severe
```

Reply:
> maybe boxing the U256 could help?

Might, definitely worth trying. Is using a `u128` an option?

Reply:
It seems that boxing here does not solve the performance issue. At least, is not clear for me, bench reporting is something tricky and I want to understand better its variability.

Since is not something really important, I put an explanation in the error description, and I revert to usize.

Reply:
Yeah, sorry. I just checked that using `U256` here doesn't affect the `std::mem::size_of::<Error>()`, so the regression must be coming from elsewhere (like `find_jump_and_sub_destinations`, longer match, allocation for `return_stack`).